game/world/triggers/codemaster.cpp

```diff
@@ -18,53 +18,58 @@ CodeMaster::CodeMaster(Vob* parent, World &world, const zenkit::VCodeMaster& cm,
   }
 
 void CodeMaster::onTrigger(const TriggerEvent &evt) {
-  size_t count = 0;
-  for(size_t i=0;i<keys.size();++i) {
-    if(!keys[i])
-      break;
-    ++count;
-    }
-
   for(size_t i=0;i<keys.size();++i) {
     if(slaves[i]==evt.emitter) {
-      if(ordered && (count!=i)) {
-        if(firstFalseIsFailure)
-          onFailure();
-        zeroState();
+      if(!ordered || count==i)
+        keys[i] = true;
+      else if(firstFalseIsFailure) {
+        onFailure();
         return;
         }
-      keys[i] = true;
+      ++count;
+      break;
       }
     }
-
-  for(auto i:keys)
-    if(!i) {
-      return;
-      }
-
-  zeroState();
-  TriggerEvent e(target,vobName,TriggerEvent::T_Trigger);
-  world.triggerEvent(e);
+  if(count<keys.size())
+    return;
+  if(std::find(keys.begin(),keys.end(),false)!=keys.end()) {
+    if(ordered)
+      onFailure();
+    } else {
+      onSuccess();
```

Comment:
code style: extra tab

also would be nice to have less nested 'if''s (and possibly more early-outs instead), for readability